packages/hardhat-core/test/internal/hardhat-network/provider/node.ts

```diff
@@ -728,16 +728,21 @@ describe("HardhatNode", () => {
         blockToRun: 7728449n, // this block has both EIP-2930 and EIP-1559 txs
         chainId: 5,
       },
+      {
+        networkName: "sepolia",
+        url: INFURA_URL.replace("mainnet", "sepolia"),
+        blockToRun: 3095000n, // this block is post-shanghai
+        chainId: 11155111,
+      },
     ];
 
     for (const { url, blockToRun, networkName, chainId } of forkedBlocks) {
       const remoteCommon = new Common({ chain: chainId });
-      const hardfork = remoteCommon.getHardforkByBlockNumber(blockToRun);
```

Comment:
This doesn't work for shanghai because `getHardforkByBlockNumber` needs the block timestamp to correctly detect that a block is in that hardfork. Now the hardfork is detected inside `runFullBlock` because we have fetched the block there and can pass the timestamp.

Reply:
Not sure if I follow this, but the rest LGTM